File: legacy/board-express/backend/src/events/derive.ts
/**
 * Pure event derivation. Given two snapshots of a card (previous + current)
 * compute the lifecycle events that happened between them. The watcher
 * persists what this returns; the SSE bus republishes them so the frontend
 * timeline can render live.
 *
 * Why derive instead of having the runner publish events directly: today
 * the runner only edits the card frontmatter on disk. The chokidar watcher
 * is what observes that. Reverse-engineering the events from the diff
 * keeps the runner contract simple (write the field, we'll notice). When
 * the runner gains a structured event stream later, this module becomes a
 * thin adapter rather than a reverse-engineer.
 */

export type CardEventType =
  | "discovered"
  | "status_changed"
  | "started"
  | "released"
  | "heartbeat"
  | "finished"
  | "verifier_called"
  | "cascade"
  | "merge_status_changed";

export interface CardSnapshot {
  readonly id: string;
  readonly status: string;
  readonly frontmatter: Record<string, unknown>;
  readonly mtimeMs: number;
}

export interface DerivedEvent {
  readonly cardId: string;
  readonly type: CardEventType;
  /** ISO 8601 timestamp. */
  readonly at: string;
  readonly details?: unknown;
}

function isString(v: unknown): v is string {
  return typeof v === "string" && v.length > 0;
}

function isNonEmpty(v: unknown): boolean {
  return v !== null && v !== undefined && !(typeof v === "string" && v === "");
}

/**
 * Coerce a frontmatter timestamp value to a canonical ISO string. If it's
 * not a parseable date, fall back to the snapshot's mtime.
 */
function toIso(value: unknown, fallbackMs: number): string {
  if (typeof value === "string") {
    const t = Date.parse(value);
    if (!Number.isNaN(t)) return new Date(t).toISOString();
  }
  if (value instanceof Date && !Number.isNaN(value.getTime())) {
    return value.toISOString();
  }
  return new Date(fallbackMs).toISOString();
}

function modelOf(fm: Record<string, unknown>): string | null {
  const used = fm["model_used"];
  if (isString(used)) return used;
  const planned = fm["model"];
  if (isString(planned)) return planned;
  return null;
}

export function deriveEvents(
  prev: CardSnapshot | null,
  current: CardSnapshot
): DerivedEvent[] {
  const events: DerivedEvent[] = [];
  const fm = current.frontmatter;
  const fallback = current.mtimeMs;

  if (prev === null) {
    events.push({
      cardId: current.id,
      type: "discovered",
      at: new Date(fallback).toISOString(),
      details: { status: current.status },
    });
    return events;
  }

  const prevFm = prev.frontmatter;

  if (prev.status !== current.status) {
    events.push({
      cardId: current.id,
      type: "status_changed",
      at: new Date(fallback).toISOString(),
      details: { from: prev.status, to: current.status },
    });
  }

  const prevClaimed = prevFm["claimed_by"];
  const currClaimed = fm["claimed_by"];
  if (!isNonEmpty(prevClaimed) && isNonEmpty(currClaimed)) {
    events.push({
      cardId: current.id,
      type: "started",
      at: toIso(fm["started_at"], fallback),
      details: { by: currClaimed, model: modelOf(fm) },
    });
  } else if (isNonEmpty(prevClaimed) && !isNonEmpty(currClaimed)) {
    events.push({
      cardId: current.id,
      type: "released",
      at: new Date(fallback).toISOString(),
      details: { from: prevClaimed },
    });
  }

  const prevHb = prevFm["last_heartbeat"];
  const currHb = fm["last_heartbeat"];
  if (isNonEmpty(currHb) && currHb !== prevHb) {
    events.push({
      cardId: current.id,
      type: "heartbeat",
      at: toIso(currHb, fallback),
      details: { by: currClaimed ?? null },
    });
  }

  const prevFinished = prevFm["finished_at"];
  const currFinished = fm["finished_at"];
  if (!isNonEmpty(prevFinished) && isNonEmpty(currFinished)) {
    const tokens = fm["actual_tokens"];
    events.push({
      cardId: current.id,
      type: "finished",
      at: toIso(currFinished, fallback),
      details: {
        tokens: typeof tokens === "number" ? tokens : null,
        model: modelOf(fm),
      },
    });
  }

  const prevVerified = prevFm["verified_at"];
  const currVerified = fm["verified_at"];
  if (!isNonEmpty(prevVerified) && isNonEmpty(currVerified)) {
    events.push({
      cardId: current.id,
      type: "verifier_called",
      at: toIso(currVerified, fallback),
      details: { by: fm["verified_by"] ?? null },
    });
  }

  // cascade_history is an append-only array. Emit one event per new entry.
  const prevCascade = Array.isArray(prevFm["cascade_history"])
    ? (prevFm["cascade_history"] as unknown[])
    : [];
  const currCascade = Array.isArray(fm["cascade_history"])
    ? (fm["cascade_history"] as unknown[])
    : [];
  if (currCascade.length > prevCascade.length) {
    for (let i = prevCascade.length; i < currCascade.length; i++) {
      const entry = currCascade[i];
      const at =
        entry && typeof entry === "object" && "at" in entry
          ? toIso((entry as { at: unknown }).at, fallback)
          : new Date(fallback).toISOString();
      events.push({
        cardId: current.id,
        type: "cascade",
        at,
        details: entry,
      });
    }
  }

  const prevMerge = prevFm["merge_status"];
  const currMerge = fm["merge_status"];
  if (isNonEmpty(currMerge) && currMerge !== prevMerge) {
    events.push({
      cardId: current.id,
      type: "merge_status_changed",
      at: new Date(fallback).toISOString(),
      details: { from: prevMerge ?? null, to: currMerge },
    });
  }

  return events;
}

File: backend/cards_api/derive.py
"""Card-event derivation — the Python port of legacy ``events/derive.ts``.

Pure function: diff the previous card snapshot against the current one and
return the events that happened. The event ``type`` strings and details
shapes are wire-contract (the Timeline UI renders them); see
docs/board/CARDS_API_CONTRACT.md §"Card events".

Timestamp rule (as legacy): an event whose trigger field carries a parseable
timestamp uses it; everything else falls back to the card's modification
time. One deliberate clarification vs legacy: for a brand-new card
(``prev is None``) we emit ``discovered`` and then apply the remaining rules
against an empty previous frontmatter, so a card that arrives already
claimed/finished still gets its lifecycle events.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

FrontMatter = dict[str, Any]


@dataclass(frozen=True)
class CardSnapshot:
    status: str
    frontmatter: FrontMatter


@dataclass(frozen=True)
class DerivedEvent:
    type: str
    at: datetime
    details: Any = field(default=None)


def _nonempty(value: Any) -> bool:
    if value is None:
        return False
    if isinstance(value, str):
        return value.strip() != ""
    return bool(value)


def _parse_ts(value: Any) -> datetime | None:
    if not isinstance(value, str) or not value.strip():
        return None
    raw = value.strip().replace("Z", "+00:00")
    try:
        parsed = datetime.fromisoformat(raw)
    except ValueError:
        return None
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=timezone.utc)
    return parsed


def _at(fm: FrontMatter, key: str, fallback: datetime) -> datetime:
    return _parse_ts(fm.get(key)) or fallback


def derive_events(
    prev: CardSnapshot | None,
    current: CardSnapshot,
    fallback_at: datetime,
) -> list[DerivedEvent]:
    events: list[DerivedEvent] = []
    cur = current.frontmatter

    if prev is None:
        events.append(
            DerivedEvent("discovered", fallback_at, {"status": current.status})
        )
        old = CardSnapshot(status=current.status, frontmatter={})
    else:
        old = prev
        if old.status != current.status:
            events.append(
                DerivedEvent(
                    "status_changed",
                    fallback_at,
                    {"from": old.status, "to": current.status},
                )
            )
    p = old.frontmatter

    if not _nonempty(p.get("claimed_by")) and _nonempty(cur.get("claimed_by")):
        events.append(
            DerivedEvent(
                "started",
                _at(cur, "claimed_at", fallback_at),
                {
                    "by": cur.get("claimed_by"),
                    "model": cur.get("model_used") or cur.get("model"),
                },
            )
        )
    if _nonempty(p.get("claimed_by")) and not _nonempty(cur.get("claimed_by")):
        events.append(
            DerivedEvent("released", fallback_at, {"from": p.get("claimed_by")})
        )

    hb = cur.get("last_heartbeat")
    if _nonempty(hb) and hb != p.get("last_heartbeat"):
        events.append(
            DerivedEvent(
                "heartbeat",
                _at(cur, "last_heartbeat", fallback_at),
                {"by": cur.get("claimed_by")},
            )
        )

    if not _nonempty(p.get("finished_at")) and _nonempty(cur.get("finished_at")):
        tokens = cur.get("actual_tokens")
        events.append(
            DerivedEvent(
                "finished",
                _at(cur, "finished_at", fallback_at),
                {
                    "tokens": tokens if isinstance(tokens, (int, float)) else None,
                    "model": cur.get("model_used") or cur.get("model"),
                },
            )
        )

    if not _nonempty(p.get("verified_at")) and _nonempty(cur.get("verified_at")):
        events.append(
            DerivedEvent(
                "verifier_called",
                _at(cur, "verified_at", fallback_at),
                {"by": cur.get("verified_by")},
            )
        )

    prev_cascade = p.get("cascade_history")
    cur_cascade = cur.get("cascade_history")
    prev_len = len(prev_cascade) if isinstance(prev_cascade, list) else 0
    if isinstance(cur_cascade, list) and len(cur_cascade) > prev_len:
        for entry in cur_cascade[prev_len:]:
            at = fallback_at
            if isinstance(entry, dict):
                at = _parse_ts(entry.get("at")) or fallback_at
            events.append(DerivedEvent("cascade", at, entry))

    merge_status = cur.get("merge_status")
    if _nonempty(merge_status) and merge_status != p.get("merge_status"):
        events.append(
            DerivedEvent(
                "merge_status_changed",
                fallback_at,
                {"from": p.get("merge_status"), "to": merge_status},
            )
        )

    return events
